ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaSource.java

```diff
@@ -0,0 +1,210 @@
+    cuUsage = new HashMap<>();
+    usageRefresher = Utils.newScheduler(1, REFRESHER_THREAD_NAME_PREFIX, true);
+    aggregationWindowsInSecs = quotaConfig.cuQuotaAggregationWindowInSecs;
+    isReady = new AtomicBoolean(false);
+  }
+
+  @Override
+  public synchronized void init() {
+    if (isReady.get()) {
+      LOGGER.warn("InMemoryCuQuotaSource is already initialized.");
+      return;
+    }
+    feUsage = new CapacityUnit();
+    cuQuota.keySet().forEach(key -> cuUsage.put(key, new CapacityUnit()));
+    usageRefresher.scheduleAtFixedRate(this::resetQuotaUsage, aggregationWindowsInSecs, aggregationWindowsInSecs,
+        TimeUnit.SECONDS);
+    isReady.set(true);
+  }
+
+  @Override
+  public boolean isReady() {
+    return isReady.get();
+  }
+
+  @Override
+  public Quota getQuota(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+
+    String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    return new Quota<>(quotaName, cuQuota.get(resourceId).getQuotaValue(quotaName), quotaResource);
+  }
+
+  @Override
+  public float getUsage(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+    String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    double usage = 0;
+    if (cuUsage.containsKey(resourceId)) {
+      usage = cuUsage.get(resourceId).getQuotaValue(quotaName);
+    }
+    return QuotaUtils.getUsagePercentage(cuQuota.get(resourceId).getQuotaValue(quotaName), usage);
+  }
+
+  @Override
+  public void chargeUsage(QuotaResource quotaResource, QuotaName quotaName, double usageCost) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+    String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    if (quotaName == QuotaName.READ_CAPACITY_UNIT) {
+      cuUsage.get(resourceId).incrementRcu((long) Math.ceil(usageCost));
+    } else {
+      cuUsage.get(resourceId).incrementWcu((long) Math.ceil(usageCost));
+    }
+  }
+
+  @Override
+  public float getSystemResourceUsage(QuotaName quotaName) throws QuotaException {
+    return QuotaUtils.getUsagePercentage(feQuota.getQuotaValue(quotaName), feUsage.getQuotaValue(quotaName));
+  }
+
+  @Override
+  public void chargeSystemResourceUsage(QuotaName quotaName, double usageCost) throws QuotaException {
+    if (quotaName == QuotaName.READ_CAPACITY_UNIT) {
+      feUsage.incrementRcu((long) Math.ceil(usageCost));
+    } else {
+      feUsage.incrementWcu((long) Math.ceil(usageCost));
+    }
+  }
+
+  @Override
+  public void updateNewQuotaResources(Collection<Account> accounts) {
+    Collection<QuotaResource> quotaResources = QuotaUtils.getQuotaResourcesFromAccounts(accounts);
+    quotaResources.forEach(quotaResource -> {
+      cuQuota.putIfAbsent(quotaResource.getResourceId(),
```

Comment:
Is `cuQuota` written/read from multiple threads? You may need a ConcurrentHashMap to prevent things like ConcurrentModificationException during iterating and, if there are multiple writers, corrupted internal state.

Reply:
Fixed the concurrency issue. Also made feUsage atomic to avoid corruption.